linker/linker_soinfo.h

```diff
@@ -338,6 +340,40 @@ struct soinfo {
   uintptr_t handle_;
 
   friend soinfo* get_libdl_info(const char* linker_path, const link_map& linker_map);
+
+ public:
+  struct SegmentInfo {
+    ElfW(Addr) phdr_addr, real_addr;
+    ElfW(Word) real_size, page_size;
+    size_t index;
+  };
+
+  typedef std::vector<SegmentInfo> seginfo_list_t;
+
+ public:
+  seginfo_list_t rand_addr_segments;
+
+  void sort_rand_addr_segments();
+
+  // Find and return a pagerando randomly mapped segment containing the given
+  // virtual address
+  SegmentInfo* find_rand_segment_info(ElfW(Addr) addr) {
```

Comment:
On another note, you might want to avoid an overflow and replace `mem_vaddr < (bin_start + seg_info.real_size)` with `(mem_vaddr - bin_start) < seg_info.real_size`